### PR TITLE
README: improve existing file

### PR DESCRIPTION
**Pull request by dougsland**
Add more description and steps to users.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,3 +2,163 @@
...
+An initramfs (Initial RAM File System) image is a fundamental component in preparing Linux systems during the boot process, preceding the initiation of the init process. 
+
+Typically, generating an initramfs involves assembling all available kernel modules and necessary files to boot and support any hardware using the specific Linux kernel version XYZ.  
+
+However, this conventional approach presents a significant challenge: loading such a voluminous image into memory during boot is time-consuming and can be problematic in critical scenarios such as healthcare, automotive, and aviation.
+
+Conversely, the initoverlayfs approach proposes a solution: "**dividing the initramfs image into two parts**." 
+
+This division entails segregating the initramfs image into two distinct components.
+The first component houses only the kernel modules necessary for storage, responsible for bringing up the operating system. Subsequently, it triggers the second component, containing all additional kernel modules and essential files required to support the Linux boot process.
```
Comment (rhatdan) [inline README.md:23]: Second part of this should be 
The second componant is mounted by the first comportment ...

File: README.md [hunk 2]
```diff
@@ -2,3 +2,163 @@
...
+    * [Step 3 - Reboot to test](#step-3---reboot-to-test)
+    * [Step 4 - Validating the boot](#step-4---validating-the-boot)
+
+# Why use initoverlayfs to generate initramfs images?
+
+An initramfs (Initial RAM File System) image is a fundamental component in preparing Linux systems during the boot process, preceding the initiation of the init process. 
+
+Typically, generating an initramfs involves assembling all available kernel modules and necessary files to boot and support any hardware using the specific Linux kernel version XYZ.  
+
+However, this conventional approach presents a significant challenge: loading such a voluminous image into memory during boot is time-consuming and can be problematic in critical scenarios such as healthcare, automotive, and aviation.
```
Comment (rhatdan) [inline README.md:18]: problematic in critical scenarios where time to boot is critical.  Edge devices commonly need to boot as quickly as possible such as healthcare, automotive and aviation.

File: README.md [hunk 3]
```diff
@@ -2,3 +2,163 @@
 
 An initial scalable filesystem for Linux operating systems
```
Comment (rhatdan) [inline README.md:3]: We should rework this section with a quick introduction to what initoverlayfs is.

File: README.md [hunk 4]
```diff
@@ -2,3 +2,163 @@
...
+# dracut -f
+# du -sh /boot/initramfs-6.5.5-300.fc39.x86_64.img
+36M	/boot/initramfs-6.5.5-300.fc39.x86_64.img
+```
+
+**Using initoverlayfs**:
+``` bash
+# /usr/bin/initoverlayfs-install
+# du -sh /boot/initramfs-6.5.5-300.fc39.x86_64.img
+13M	/boot/initramfs-6.5.5-300.fc39.x86_64.img
```
Comment (ericcurtin) [inline README.md:40]: @masneyb is looking into things to make this even smaller. Eventually we will likely be able to remove systemd also, leaving all systemd stuff in initoverlayfs. Note the goal is not to rewrite systemd, we just need some things to spin up storage quickly and then switch to systemd (or alternative init process, none of this solution is systemd specific).

Eventually I expect this to be single-digit M, like < 10M.

File: README.md [hunk 5]
```diff
@@ -2,3 +2,163 @@
...
+Oct 25 00:20:53 dell730.medogz.com storage-init: (stat("/initoverlayfs/dev", 0x7ffe85ccaa10) == 0) && 5 != 18)
+Oct 25 00:20:53 dell730.medogz.com storage-init: mount("/boot", "/initoverlayfs/boot", "ext4", MS_MOVE, NULL) 2 (No such file or directory)
+Oct 25 00:20:53 dell730.medogz.com storage-init: forked 368 fork_execlp
+Oct 25 00:20:53 dell730.medogz.com storage-init: fork_execlp("udevadm")
+Oct 25 00:20:53 dell730.medogz.com storage-init: forked 357 fork_execvp_no_wait
+Oct 25 00:20:53 dell730.medogz.com storage-init: fork_execvp_no_wait(0x16c4840)
+Oct 25 00:20:53 dell730.medogz.com storage-init: bootfs: {"UUID=b7eaec82-7c35-4887-becf-60ee7889624f", "bootfs UUID=b7eaec82-7c35-4887-becf-60ee7889624f"}, bootfstype: {"ext4", "bootfstype ext4"}, fs: {"(null)", "(null)"}, fstype: {"(null)", "(null)"}, udev_trigger: {"udevadm trigger --type=devices --action=add --subsystem-match=module --subsystem-match=block --subsystem-match=virtio --subsystem-match=pci --subsystem-match=nvme", "udev_trigger udevadm trigger --type=devices --action=add --subsystem-match=module --subsystem-match=block --subsystem-match=virtio --subsystem-match=pci --subsystem-match=nvme"}
```
Comment (ericcurtin) [inline README.md:161]: Ah you found my verbose logging :smile: ... Eventually I will turn many of these logs off (or make it configurable), it was useful at the start to get it working...

**Comment by ericcurtin** [comment]
I will do a detailed review of this in the next few hours, but thanks so much for this, on first initial skim the majority looks great!

File: README.md (hunk 1 above)
Comment (ericcurtin) [reply README.md:23]: I would say "The first component (initramfs) houses only the kernel modules and udev-rules necessary for storage, responsible for bringing up storage device quickly. Subsequently, it mounts and switches the second component (initoverlayfs), containing all additional kernel modules and essential files required to support the Linux boot process."

The first part does storage and nothing else. There's actually a third component rootfs, but nothing there changes I guess. Some kernel modules are in rootfs and not initoverlayfs, just as some kernel modules will be in initoverlayfs and not initramfs.

There is a gluttony of essential files put in initoverlayfs that would have been in initramfs in the past, such as ostree tools, decryption tools, filesystem verification tools, bootsplash tools, it's 1,000s of files, but "essential files" is good enough without delving into the detail like you did. :+1:

**Comment by ericcurtin** [comment]
One thing worth adding, is we would regard this in alpha or beta stage, we've only tested on three types of storage device so far:

- virtio
- NVMe
- Qualcomm UFS specifically (there's a qualcomm specific driver for this)

Although it may work on many of the other storage devices, I can't guarantee it yet. Don't want to be on the hook for helping debug unbootable machines. It's often easy to fix, but prevention right now is better than curing!

**Comment by dougsland** [comment]
> One thing worth adding, is we would regard this in alpha or beta stage, we've only tested on three types of storage device so far:
> 
> * virtio
> * NVMe
> * Qualcomm UFS specifically (there's a qualcomm specific driver for this)
> 
> Although it may work on other kinds of storage devices not covered here, I can't guarantee it yet. Don't want to be on the hook for helping debug unbootable machines. It's often easy to fix, but prevention right now is better than curing!

Sure thing @ericcurtin. I am just waiting all feedback from you and @rhatdan to update this PR.

**Comment by ericcurtin** [comment]
No more feedback on my end @dougsland :smile:

File: README.md (hunk 4 above)
Comment (dougsland) [reply README.md:40]: This will be very nice!

File: README.md (hunk 3 above)
Comment (dougsland) [reply README.md]: @rhatdan @ericcurtin added a short sentence, please review!

**Comment by rhatdan** [comment]
LGTM

**Comment by ericcurtin** [comment]
More contributors :champagne: 

I'm gonna expand this even more sometime to explain the scalability, maybe it's just me but I think that is just as important that the raw boot performance. Before this, it was hard to even consider using meatier technologies this early in boot because of the scalability issue of initramfs.

**Comment by rhatdan** [comment]
I agree, that was some of the stuff I was thinking of, now the initramfs can be huge, stick python and friends in there, we don't care.

**Comment by dougsland** [comment]
Thanks for the merge, more patches to come.

**Pull request merged**